validator/README.md

```diff
@@ -19,30 +19,62 @@ limitations under the License.
 A validator for the
 [AMP HTML format](https://github.com/ampproject/amphtml/blob/master/README.md).
 
+To use, install Node.js version 4.X on your system (tested with v4.4.2). E.g.,
+[by downloading](https://nodejs.org/en/download/) or
+[by using a package manager](https://nodejs.org/en/download/package-manager/) or
+[by using NVM](https://github.com/creationix/nvm).
+
 ## Using the command-line tool
 
-* Install Node.js version 4.X on your system. E.g.,
-  [by downloading](https://nodejs.org/en/download/) or
-  [by using a package manager](https://nodejs.org/en/download/package-manager/) or
-  [by using NVM](https://github.com/creationix/nvm).
-* Type `node index.js` in this directory to get started:
+* Type `./index.js` in this directory to get started:
 
 ```
-$ node index.js testdata/feature_tests/minimum_valid_amp.html
+$ ./index.js testdata/feature_tests/minimum_valid_amp.html
 testdata/feature_tests/minimum_valid_amp.html: PASS
 
-$ node index.js testdata/feature_tests/several_errors.html
+$ ./index.js testdata/feature_tests/several_errors.html
 testdata/feature_tests/several_errors.html:23:2 The attribute 'charset' may not appear in tag 'meta name= and content='.
 testdata/feature_tests/several_errors.html:26:2 The tag 'script' is disallowed except in specific forms.
 testdata/feature_tests/several_errors.html:32:2 The mandatory attribute 'height' is missing in tag 'amp-img'. (see https://www.ampproject.org/docs/reference/amp-img.html)
 testdata/feature_tests/several_errors.html:34:2 The attribute 'width' in tag 'amp-ad' is set to the invalid value '100%'. (see https://www.ampproject.org/docs/reference/amp-ad.html)
 ...
 ```
 
-* If you wish to install the Validator as a system command,
-  install the NPM package manager (e.g. using apt-get in Ubuntu Linux) and
-  run `npm install -g` in this directory. After that, you may type
-  `amp_validator` in any directory to invoke the validator.
+If you wish to install the Validator as a system command,
+install the NPM package manager (e.g. using apt-get in Ubuntu Linux) and
+run `npm install -g` in this directory. After that, you may type
+`amp-validator` in any directory to invoke the validator.
+
+## Using the web UI
+
+This is a simple web editor which validates AMP HTML documents on the fly.
+This feature is new and experimental, feedback is especially welcome.
+
+* Type `./index.js webui` in this directory.
+* Point your web browser to http://127.0.0.1:8765/
+
+## Using the NodeJS API
+
+This API is new and still experimental, feedback is especially welcome. We may
+try to port to earlier versions of NodeJS if sufficient interest exists.
+
+```
+'use strict';
+
+const ampValidator = require('amp-validator');
+ampValidator.validateString('<html>Hello, world</html>').then(
+    (result) => {
+        ((result.status === 'PASS') ? console.log : console.error)(result.status);
```

Comment:
I think this would have the problem that the promise could either be rejected due to a problem instantiating the validator (e.g., no connection to www.ampproject.org where the script usually comes from) or something wrong with the page - I think this could get confusing. Further, I think there is this problem that we have a record that's called ValidationResult which then contains errors, some of which can be warnings, but now we'd need to make ValidationResult either extend Error or not depending on whether it's passing.
Maybe a better way is to detangle obtaining a validator instance from validateString. The former could be promised based, whereas the latter would just return an object.
```js
const ampValidator = require('amp-validator');
ampValidator.getInstance().then((validator) => {
   const validationResult = validator.validateString('<html>Hello, world</html>');
   ((result.status === 'PASS') ? console.log : console.error)(result.status);
   for (const error of result.errors) {
     let msg = 'line ' + error.line + ', col ' + error.col + ': ' + error.message;
     if (error.specUrl !== null) {
       msg += ' (see ' + error.specUrl + ')';
     }
     ((error.severity === 'ERROR') ? console.error : console.warn)(msg);
   }
}
```
The getInstance function is then also where a user could supply a parameter in case they wanted to use a locally built or fetched validator script. But validateString would just return a ValidationResult object, which neatly corresponds to the proto buffer in
https://github.com/powdercloud/amphtml/blob/83277a14dd79794d19a10500db22d1460e3a1644/validator/validator.proto
I'll try this and see how it straightens out the code. If there's a better way than getInstance, happy for suggestions - including if you think this is no better than the promise-based validateString. My hunch is that this has the edge if we end up adding additional validator functions.